### DISCUSSION TITLE
Textarea with auto-resize gets reset after wire:model request

### DISCUSSION THREAD

**Discussion by n0ah01** [livewire/livewire#7066] [1. Help] [answered]
I have a livewire-component with a textarea, that has some javascript to automatically resize the textarea to always display all the text in the textarea. Initially everything is working fine until livewire saves the new value, where the textarea is reset to the initial size creating an overflow (scroll). It starts to work again when I type someting new into the textarea, until the value is again sent to the server.

```HTML
<textarea
	class="flex-1 w-full min-h-[10px] resize-none"
	x-init="$nextTick(() => resize())"
	x-data="{
		resize() {
			$el.style.height = '0px';
			$el.style.height = $el.scrollHeight + 'px';
		}
	}"
	wire:model.blur="name"
	@input="resize()"></textarea>
```

How can I get my textarea to always display its value fully?

I understand that a component is "replaced" when an update happens, and the x-init function is not rerun because it is not a new component/pageload. I also tried to use alpine.js's x-model to link the value hopping that would prevent the "replacement", without luck. Is there some event that I can listen to to rerun the resize() function after livewire did its update?

Thank you guys in advance!

**Comment by n0ah01** [comment, marked as answer]
Made a workaround. It has multiple fronts, maybe it'll help someone else:

On the component where the textarea is located I have a _show_ livewire-variable that controlls the x-show of the component. There is a $watch() on the _show_ that triggers the resize-function that resizes the textarea. So the textarea is resized after when the component is shown/filled with data:
```
x-init="$watch('show', (val) => {
if (val) {
	$nextTick(() => {
		resize();
	});
}
})"
x-data="{
show: @entangle('show').live,
titleHeight: 10,
resize() {
	this.titleHeight = $refs.kanbanDetailTitle.scrollHeight + 2;
}
}"
x-show="show"
wire:submit="save_fn()"
@save_fn="resize()"
```

Then the save-function has the renderless attribute, to prevent the rerendering of my component.
```
use Livewire\Attributes\Renderless;
...
#[Renderless]
public function save_fn() {
	// saving...
}
```

I'm sure there are prettier ways of achiving my goal, but for the moment it works.

**Comment by FranciscoDumont** [comment]
I faced the same issue with textarea resizing and found a solution that worked for me. 
I want to share it here, though I acknowledge it might not be the most elegant one due to the use of `setTimeout`.

I added a new event `x-on:resize-textarea.window` that listens for the custom event `resize-textarea` on the window.
Within this event, I used `setTimeout` to delay the execution of the `resize()` function after a short period of time.

Blade View:
```html
<textarea id="descriptionTextarea"
          wire:ignore
          wire:model.blur="budgetDescription"
          x-init="$nextTick(() => resize())"
          x-data="{
                    resize() {
                      $el.style.height = '0px';
                      $el.style.height = $el.scrollHeight + 20 + 'px';
                    }
                  }"
          @input="resize()"
          
          <!-- I added this part-->
          x-on:resize-textarea.window="
                    setTimeout(() => {
                        resize();
                        }, 100);
                    "
                    
          class="form-control" placeholder="Descripción de la propuesta" ></textarea>

```

Livewire controller:
``` php
public function render()
{
    $this->dispatch('resize-textarea');
    return view('livewire.budget-maker');
}

```

While this workaround does the job, I want to note that I had to use `setTimeout` as it was the only way I found to make it work. Without it, the solution wasn't functioning as expected.

I'm sharing this here in case it helps, and I'd love to hear if anyone has suggestions for improving this approach.

Feel free to share your thoughts or suggest alternative solutions! 🚀

**Comment by Valkje** [comment]
Maybe a bit late to the party but I ran into the same issue with Livewire 3. My solution builds on the ones suggested above but makes use of Livewire's `rendered` hook, where I dispatch a `'post-rendered'` event:

```php
class EditPost extends Component
{
    // Code here...

    // Executes after render is called
    public function rendered($view, $html)
    {
        $this->dispatch('post-rendered');
    }
}
```

Then in the view, I don't only resize on init and on input entry, but also when the `'post-rendered'` event fires:

```html
<textarea
    x-init="$nextTick(resize); $wire.on('post-rendered', () => $nextTick(resize))"
    x-data="{
        resize() {
            $el.style.height = '0px';
            $el.style.height = $el.scrollHeight + 'px';
        }
    }"
    x-on:input="resize()"
    wire:model.live.throttle.500ms="post.body"
></textarea>
```

I am still not 100% happy with the result as it flickers every time the component `$post->body` updates (in this case every 500ms), but for now it works well enough. Interested to see any improvements!

**Reply by timsteinhauer-bow** [reply]
Hi, i moved the inline styles to a style-tag and its working great.

```
<div class="input-group mb-3 {{ $config["class"] ?? "" }}"
     style="{{ $config["style"] ?? "" }}"
>
    @php( $cssIdSelector = '__'.str_replace(".", "_", $key))
    @php( $cssId = str_replace(".", "_", $key))
 
    <div class="input-group-prepend">
        <span class="input-group-text">{{ $title }}</span>
    </div>
   
    <textarea
        id="{{ $cssId }}"
        class="form-control @error($key) is-invalid @enderror"
        wire:model="{{ $keyPath }}"
       
        x-data="{ resize: () => {
            document.getElementById('{{ $cssIdSelector }}').innerHTML
            = '#{{ $cssId }}{height: 35px;}';
            document.getElementById('{{ $cssIdSelector }}').innerHTML
            = '#{{ $cssId }}{height:' + ($el.scrollHeight + 2) + 'px;}'
                } }"
        x-init="resize()"
       @input="resize()"
    ></textarea>
       
    <style wire:ignore id="{{ $cssIdSelector }}"></style>
</div>

```
